test/NuGet.Clients.Tests/NuGet.VisualStudio.Common.Test/Telemetry/PackageSourceTelemetryTests.cs

```diff
@@ -104,38 +120,76 @@ public void AddAggregateData_TimingAggregation()
+            Assert.Equal(timings.Length, http.Requests);
+            Assert.Equal(timings.Aggregate(TimeSpan.Zero, (a, b) => a + b), http.TotalDuration);
         }
 
         [Fact]
-        public void AddAggregateData_IsThreadSafe()
+        public void AddNupkgCopiedData_MultipleEvents_AccumulatesCorrectly()
         {
             // Arrange
+            var sizes = new[] { 1, 22, 333, 4444, 55555, 666666 };
             var data = new ConcurrentDictionary<string, PackageSourceTelemetry.Data>();
-            var eventsToRaise = 10000;
 
             // Act
-            Parallel.For(0, eventsToRaise, _ =>
+            for (int i = 0; i < sizes.Length; i++)
+            {
+                var nce = new ProtocolDiagnosticNupkgCopiedEvent("source", sizes[i]);
+                PackageSourceTelemetry.AddNupkgCopiedData(nce, data);
+            }
+
+            // Assert
+            var result = Assert.Single(data).Value;
+            Assert.Equal(sizes.Length, result.NupkgCount);
+            Assert.Equal(sizes.Sum(), result.NupkgSize);
+        }
+
+        [Fact]
+        public async Task AddData_IsThreadSafe()
+        {
+            // Arrange
+            var data = new ConcurrentDictionary<string, PackageSourceTelemetry.Data>();
+            var stringTable = new ConcurrentDictionary<string, ConcurrentDictionary<string, string>>();
+            var eventsToRaise = 10000;
+
+            Task SendEvents(Action action)
             {
-                var pde = CreateSampleProtocolDiagnosticsEvent();
-                PackageSourceTelemetry.AddAggregateData(pde, data);
-            });
+                var tasks = new List<Task>();
+
+                for (int i =0; i < eventsToRaise; i++)
+                {
+                    tasks.Add(Task.Run(() => action()));
+                }
+
+                return Task.WhenAll(tasks);
+            }
+            var source = "http://source.test/v3/index.json";
+            var resourceEvent = CreateSampleResourceEvent(source: source);
+            var httpEvent = CreateSampleHttpEvent(source: source);
+            var nupkgCopiedEvent = new ProtocolDiagnosticNupkgCopiedEvent(source, fileSize: 123456);
+
+            // Act
+            var resourceEvents = Task.Run(() => SendEvents(() => PackageSourceTelemetry.AddResourceData(resourceEvent, data, stringTable)));
+            var httpEvents = Task.Run(() => SendEvents(() => PackageSourceTelemetry.AddHttpData(httpEvent, data)));
+            var nupkgCopiedEvents = Task.Run(() => SendEvents(() => PackageSourceTelemetry.AddNupkgCopiedData(nupkgCopiedEvent, data)));
+            await Task.WhenAll(resourceEvents, httpEvents, nupkgCopiedEvents);
 
             // Assert
             KeyValuePair<string, PackageSourceTelemetry.Data> pair = Assert.Single(data);
-            Assert.Equal(eventsToRaise, pair.Value.Nupkg.EventTiming.Requests);
+            Assert.Equal(eventsToRaise, pair.Value.Resources.Sum(r => r.Value.count));
```

Comment:
Not following how we are comparing the constant, 10,000, to the Sum of Resource's value's count.
Seems the more interesting test would be that the number of bytes (total nupkg size) got added up correctly.

Reply:
discussed offline